File: pulse-kotlin/src/commonMain/kotlin/com/fluxchess/pulse/kotlin/engine/MoveGenerator.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.engine

fun generateLegalMoves(moveList: MoveList, position: Position) {
	generateMoves(moveList, position)

	val size = moveList.size
	moveList.reset()
	for (i in 0 until size) {
		val move = moveList.entries[i].move
		position.makeMove(move)
		if (!position.isCheck(oppositeOf(position.activeColor))) {
			moveList.add(move)
		}
		position.undoMove(move)
	}
}

fun generateMoves(moveList: MoveList, position: Position) {
	moveList.reset()
	addAllMoves(moveList, position)
	if (!position.isCheck()) {
		val square = next(position.pieces[position.activeColor][KING])
		addCastlingMoves(moveList, position, square)
	}

	moveList.rateByMVVLVA()
	moveList.sort()
}

fun generateQuiescentMoves(moveList: MoveList, position: Position) {
	moveList.reset()
	addAllMoves(moveList, position)
	if (!position.isCheck()) {
		val size = moveList.size
		moveList.reset()
		for (i in 0 until size) {
			if (targetPieceOf(moveList.entries[i].move) != NO_PIECE) {
				moveList.add(moveList.entries[i].move)
			}
		}
	}

	moveList.rateByMVVLVA()
	moveList.sort()
}

private fun addAllMoves(moveList: MoveList, position: Position) {
	val activeColor = position.activeColor

	var squares = position.pieces[activeColor][PAWN]
	while (squares.compareTo(0u) != 0) {
		val square = next(squares)
		addPawnMoves(moveList, position, square)
		squares = remainder(squares)
	}
	squares = position.pieces[activeColor][KNIGHT]
	while (squares.compareTo(0u) != 0) {
		val square = next(squares)
		addPieceMoves(moveList, position, square, knightDirections)
		squares = remainder(squares)
	}
	squares = position.pieces[activeColor][BISHOP]
	while (squares.compareTo(0u) != 0) {
		val square = next(squares)
		addPieceMoves(moveList, position, square, bishopDirections)
		squares = remainder(squares)
	}
	squares = position.pieces[activeColor][ROOK]
	while (squares.compareTo(0u) != 0) {
		val square = next(squares)
		addPieceMoves(moveList, position, square, rookDirections)
		squares = remainder(squares)
	}
	squares = position.pieces[activeColor][QUEEN]
	while (squares.compareTo(0u) != 0) {
		val square = next(squares)
		addPieceMoves(moveList, position, square, queenDirections)
		squares = remainder(squares)
	}
	val square = next(position.pieces[activeColor][KING])
	addPieceMoves(moveList, position, square, kingDirections)
}

private fun addPawnMoves(moveList: MoveList, position: Position, pawnSquare: Square) {
	val pawnPiece = position.board[pawnSquare]
	val pawnColor = pieceColorOf(pawnPiece)

	for (direction in pawnCapturingDirections[pawnColor]) {
		val targetSquare = pawnSquare + direction
		if (isValidSquare(targetSquare)) {
			var targetPiece = position.board[targetSquare]
			if (targetPiece != NO_PIECE) {
				if (pieceColorOf(targetPiece) == oppositeOf(pawnColor)) {
					// Capturing move
					if ((pawnColor == WHITE && rankOf(targetSquare) == RANK_8)
						|| (pawnColor == BLACK && rankOf(targetSquare) == RANK_1)
					) {
						// Pawn promotion capturing move
						moveList.add(
							moveOf(PAWN_PROMOTION_MOVE, pawnSquare, targetSquare, pawnPiece, targetPiece, QUEEN),
						)
						moveList.add(
							moveOf(PAWN_PROMOTION_MOVE, pawnSquare, targetSquare, pawnPiece, targetPiece, ROOK),
						)
						moveList.add(
							moveOf(PAWN_PROMOTION_MOVE, pawnSquare, targetSquare, pawnPiece, targetPiece, BISHOP),
						)
						moveList.add(
							moveOf(PAWN_PROMOTION_MOVE, pawnSquare, targetSquare, pawnPiece, targetPiece, KNIGHT),
						)
					} else {
						// Normal capturing move
						moveList.add(
							moveOf(NORMAL_MOVE, pawnSquare, targetSquare, pawnPiece, targetPiece, NO_PIECE_TYPE),
						)
					}
				}
			} else if (targetSquare == position.enPassantSquare) {
				// En passant move
				val oppositeDirection = pawnMoveDirections[oppositeOf(pawnColor)]
				val captureSquare = targetSquare + oppositeDirection
				targetPiece = position.board[captureSquare]
				moveList.add(moveOf(EN_PASSANT_MOVE, pawnSquare, targetSquare, pawnPiece, targetPiece, NO_PIECE_TYPE))
			}
		}
	}

	val direction = pawnMoveDirections[pawnColor]

	// Move one rank forward
	var targetSquare = pawnSquare + direction
	if (isValidSquare(targetSquare) && position.board[targetSquare] == NO_PIECE) {
		if ((pawnColor == WHITE && rankOf(targetSquare) == RANK_8)
			|| (pawnColor == BLACK && rankOf(targetSquare) == RANK_1)
		) {
			// Pawn promotion move
			moveList.add(moveOf(PAWN_PROMOTION_MOVE, pawnSquare, targetSquare, pawnPiece, NO_PIECE, QUEEN))
			moveList.add(moveOf(PAWN_PROMOTION_MOVE, pawnSquare, targetSquare, pawnPiece, NO_PIECE, ROOK))
			moveList.add(moveOf(PAWN_PROMOTION_MOVE, pawnSquare, targetSquare, pawnPiece, NO_PIECE, BISHOP))
			moveList.add(moveOf(PAWN_PROMOTION_MOVE, pawnSquare, targetSquare, pawnPiece, NO_PIECE, KNIGHT))
		} else {
			// Normal move
			moveList.add(moveOf(NORMAL_MOVE, pawnSquare, targetSquare, pawnPiece, NO_PIECE, NO_PIECE_TYPE))

			targetSquare += direction
			if (isValidSquare(targetSquare) && position.board[targetSquare] == NO_PIECE) {
				if ((pawnColor == WHITE && rankOf(targetSquare) == RANK_4)
					|| (pawnColor == BLACK && rankOf(targetSquare) == RANK_5)
				) {
					// Pawn double move
					moveList.add(moveOf(PAWN_DOUBLE_MOVE, pawnSquare, targetSquare, pawnPiece, NO_PIECE, NO_PIECE_TYPE))
				}
			}
		}
	}
}

private fun addPieceMoves(moveList: MoveList, position: Position, originSquare: Square, directions: Array<Direction>) {
	val originPiece = position.board[originSquare]
	val sliding = isSliding(pieceTypeOf(originPiece))
	val oppositeColor = oppositeOf(pieceColorOf(originPiece))

	for (direction in directions) {
		var targetSquare = originSquare + direction
		while (isValidSquare(targetSquare)) {
			val targetPiece = position.board[targetSquare]
			if (targetPiece != NO_PIECE) {
				if (pieceColorOf(targetPiece) == oppositeColor) {
					// Capturing move
					moveList.add(
						moveOf(NORMAL_MOVE, originSquare, targetSquare, originPiece, targetPiece, NO_PIECE_TYPE),
					)
				}
				break
			} else {
				// Normal move
				moveList.add(moveOf(NORMAL_MOVE, originSquare, targetSquare, originPiece, NO_PIECE, NO_PIECE_TYPE))
				if (!sliding) {
					break
				}
				targetSquare += direction
			}
		}
	}
}

private fun addCastlingMoves(moveList: MoveList, position: Position, kingSquare: Square) {
	val kingPiece = position.board[kingSquare]

	if (pieceColorOf(kingPiece) == WHITE) {
		// Do not test g1 whether it is attacked as we will test it later
		if ((position.castlingRights and WHITE_KINGSIDE) != NO_CASTLING
			&& position.board[F1] == NO_PIECE
			&& position.board[G1] == NO_PIECE
			&& !position.isAttacked(F1, BLACK)
		) {
			moveList.add(moveOf(CASTLING_MOVE, kingSquare, G1, kingPiece, NO_PIECE, NO_PIECE_TYPE))
		}
		// Do not test c1 whether it is attacked as we will test it later
		if ((position.castlingRights and WHITE_QUEENSIDE) != NO_CASTLING
			&& position.board[B1] == NO_PIECE
			&& position.board[C1] == NO_PIECE
			&& position.board[D1] == NO_PIECE
			&& !position.isAttacked(D1, BLACK)
		) {
			moveList.add(moveOf(CASTLING_MOVE, kingSquare, C1, kingPiece, NO_PIECE, NO_PIECE_TYPE))
		}
	} else {
		// Do not test g8 whether it is attacked as we will test it later
		if ((position.castlingRights and BLACK_KINGSIDE) != NO_CASTLING
			&& position.board[F8] == NO_PIECE
			&& position.board[G8] == NO_PIECE
			&& !position.isAttacked(F8, WHITE)
		) {
			moveList.add(moveOf(CASTLING_MOVE, kingSquare, G8, kingPiece, NO_PIECE, NO_PIECE_TYPE))
		}
		// Do not test c8 whether it is attacked as we will test it later
		if ((position.castlingRights and BLACK_QUEENSIDE) != NO_CASTLING
			&& position.board[B8] == NO_PIECE
			&& position.board[C8] == NO_PIECE
			&& position.board[D8] == NO_PIECE
			&& !position.isAttacked(D8, WHITE)
		) {
			moveList.add(moveOf(CASTLING_MOVE, kingSquare, C8, kingPiece, NO_PIECE, NO_PIECE_TYPE))
		}
	}
}

File: pulse-go/internal/pulse/engine/move_generator.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

func GenerateLegalMoves(ml *MoveList, p *Position) {
	GenerateMoves(ml, p)

	size := ml.Size
	ml.reset()
	for i := 0; i < size; i++ {
		m := ml.Entries[i].Move
		p.MakeMove(m)
		if !p.isCheckFor(OppositeOf(p.ActiveColor)) {
			ml.add(m)
		}
		p.UndoMove(m)
	}
}

func GenerateMoves(ml *MoveList, p *Position) {
	ml.reset()
	addAllMoves(ml, p)
	if !p.isCheck() {
		sq := next(p.pieces[p.ActiveColor][King])
		addCastlingMoves(ml, p, sq)
	}

	ml.rateByMVVLVA()
	ml.sort()
}

func GenerateQuiescentMoves(ml *MoveList, p *Position) {
	ml.reset()
	addAllMoves(ml, p)
	if !p.isCheck() {
		size := ml.Size
		ml.reset()
		for i := 0; i < size; i++ {
			if TargetPieceOf(ml.Entries[i].Move) != NoPiece {
				ml.add(ml.Entries[i].Move)
			}
		}
	}

	ml.rateByMVVLVA()
	ml.sort()
}

func addAllMoves(ml *MoveList, p *Position) {
	activeCol := p.ActiveColor

	sqs := p.pieces[activeCol][Pawn]
	for sqs != 0 {
		sq := next(sqs)
		addPawnMoves(ml, p, sq)
		sqs = remainder(sqs)
	}
	sqs = p.pieces[activeCol][Knight]
	for sqs != 0 {
		sq := next(sqs)
		addPieceMoves(ml, p, sq, knightDirections[:])
		sqs = remainder(sqs)
	}
	sqs = p.pieces[activeCol][Bishop]
	for sqs != 0 {
		sq := next(sqs)
		addPieceMoves(ml, p, sq, bishopDirections[:])
		sqs = remainder(sqs)
	}
	sqs = p.pieces[activeCol][Rook]
	for sqs != 0 {
		sq := next(sqs)
		addPieceMoves(ml, p, sq, rookDirections[:])
		sqs = remainder(sqs)
	}
	sqs = p.pieces[activeCol][Queen]
	for sqs != 0 {
		sq := next(sqs)
		addPieceMoves(ml, p, sq, queenDirections[:])
		sqs = remainder(sqs)
	}
	sq := next(p.pieces[activeCol][King])
	addPieceMoves(ml, p, sq, kingDirections[:])
}

func addPawnMoves(ml *MoveList, p *Position, pawnSq Square) {
	pawnPc := p.board[pawnSq]
	pawnCol := PieceColorOf(pawnPc)

	for _, dir := range pawnCapturingDirections[pawnCol] {
		targetSq := pawnSq + dir
		if IsValidSquare(targetSq) {
			targetPc := p.board[targetSq]
			if targetPc != NoPiece {
				if PieceColorOf(targetPc) == OppositeOf(pawnCol) {
					// Capturing move
					if (pawnCol == White && RankOf(targetSq) == Rank8) ||
						(pawnCol == Black && RankOf(targetSq) == Rank1) {
						// Pawn promotion capturing move
						ml.add(moveOf(pawnPromotionMove, pawnSq, targetSq, pawnPc, targetPc, Queen))
						ml.add(moveOf(pawnPromotionMove, pawnSq, targetSq, pawnPc, targetPc, Rook))
						ml.add(moveOf(pawnPromotionMove, pawnSq, targetSq, pawnPc, targetPc, Bishop))
						ml.add(moveOf(pawnPromotionMove, pawnSq, targetSq, pawnPc, targetPc, Knight))
					} else {
						// Normal capturing move
						ml.add(moveOf(normalMove, pawnSq, targetSq, pawnPc, targetPc, NoPieceType))
					}
				}
			} else if targetSq == p.EnPassantSquare {
				// En passant move
				oppositeDir := pawnMoveDirections[OppositeOf(pawnCol)]
				captureSq := targetSq + oppositeDir
				targetPc = p.board[captureSq]
				ml.add(moveOf(enPassantMove, pawnSq, targetSq, pawnPc, targetPc, NoPieceType))
			}
		}
	}

	dir := pawnMoveDirections[pawnCol]

	// Move one rank forward
	targetSq := pawnSq + dir
	if IsValidSquare(targetSq) && p.board[targetSq] == NoPiece {
		if (pawnCol == White && RankOf(targetSq) == Rank8) ||
			(pawnCol == Black && RankOf(targetSq) == Rank1) {
			// Pawn promotion move
			ml.add(moveOf(pawnPromotionMove, pawnSq, targetSq, pawnPc, NoPiece, Queen))
			ml.add(moveOf(pawnPromotionMove, pawnSq, targetSq, pawnPc, NoPiece, Rook))
			ml.add(moveOf(pawnPromotionMove, pawnSq, targetSq, pawnPc, NoPiece, Bishop))
			ml.add(moveOf(pawnPromotionMove, pawnSq, targetSq, pawnPc, NoPiece, Knight))
		} else {
			// Normal move
			ml.add(moveOf(normalMove, pawnSq, targetSq, pawnPc, NoPiece, NoPieceType))

			targetSq += dir
			if IsValidSquare(targetSq) && p.board[targetSq] == NoPiece {
				if (pawnCol == White && RankOf(targetSq) == Rank4) ||
					(pawnCol == Black && RankOf(targetSq) == Rank5) {
					// Pawn double move
					ml.add(moveOf(pawnDoubleMove, pawnSq, targetSq, pawnPc, NoPiece, NoPieceType))
				}
			}
		}
	}
}

func addPieceMoves(ml *MoveList, p *Position, originSq Square, directions []direction) {
	originPc := p.board[originSq]
	sliding := isSliding(PieceTypeOf(originPc))
	oppositeCol := OppositeOf(PieceColorOf(originPc))

	for _, dir := range directions {
		targetSq := originSq + dir
		for IsValidSquare(targetSq) {
			targetPc := p.board[targetSq]
			if targetPc != NoPiece {
				if PieceColorOf(targetPc) == oppositeCol {
					// Capturing move
					ml.add(moveOf(normalMove, originSq, targetSq, originPc, targetPc, NoPieceType))
				}
				break
			} else {
				// Normal move
				ml.add(moveOf(normalMove, originSq, targetSq, originPc, NoPiece, NoPieceType))
				if !sliding {
					break
				}
				targetSq += dir
			}
		}
	}
}

func addCastlingMoves(ml *MoveList, p *Position, kingSq Square) {
	kingPc := p.board[kingSq]

	if PieceColorOf(kingPc) == White {
		// Do not test g1 whether it is attacked as we will test it later
		if (p.CastlingRights&WhiteKingside) != NoCastling &&
			p.board[F1] == NoPiece &&
			p.board[G1] == NoPiece &&
			!p.isAttacked(F1, Black) {
			ml.add(moveOf(castlingMove, kingSq, G1, kingPc, NoPiece, NoPieceType))
		}
		// Do not test c1 whether it is attacked as we will test it later
		if (p.CastlingRights&WhiteQueenside) != NoCastling &&
			p.board[B1] == NoPiece &&
			p.board[C1] == NoPiece &&
			p.board[D1] == NoPiece &&
			!p.isAttacked(D1, Black) {
			ml.add(moveOf(castlingMove, kingSq, C1, kingPc, NoPiece, NoPieceType))
		}
	} else {
		// Do not test g8 whether it is attacked as we will test it later
		if (p.CastlingRights&BlackKingside) != NoCastling &&
			p.board[F8] == NoPiece &&
			p.board[G8] == NoPiece &&
			!p.isAttacked(F8, White) {
			ml.add(moveOf(castlingMove, kingSq, G8, kingPc, NoPiece, NoPieceType))
		}
		// Do not test c8 whether it is attacked as we will test it later
		if (p.CastlingRights&BlackQueenside) != NoCastling &&
			p.board[B8] == NoPiece &&
			p.board[C8] == NoPiece &&
			p.board[D8] == NoPiece &&
			!p.isAttacked(D8, White) {
			ml.add(moveOf(castlingMove, kingSq, C8, kingPc, NoPiece, NoPieceType))
		}
	}
}
